state/protocol/badger/mutator_test.go

```diff
@@ -1993,6 +1993,451 @@ func TestEmergencyEpochFallback(t *testing.T) {
+
+			metricsMock.On("CurrentEpochCounter", epoch2Setup.Counter).Once()
+			metricsMock.On("EpochTransitionHeight", block8.Header.Height).Once()
+			metricsMock.On("CurrentEpochFinalView", epoch2Setup.FinalView).Once()
+			protoEventsMock.On("EpochTransition", epoch2Setup.Counter, block8.Header).Once()
+
+			// epoch transition happens at this point
+			unittest.InsertAndFinalize(t, state, block8)
+
+			metricsMock.AssertCalled(t, "CurrentEpochCounter", epoch2Setup.Counter)
+			metricsMock.AssertCalled(t, "EpochTransitionHeight", block8.Header.Height)
+			metricsMock.AssertCalled(t, "CurrentEpochFinalView", epoch2Setup.FinalView)
+			protoEventsMock.AssertCalled(t, "EpochTransition", epoch2Setup.Counter, block8.Header)
+
+			// B9 doesn't have any seals, but it reaches the safety threshold for the current epoch, meaning we will create an EpochExtension
+			block9 := unittest.BlockWithParentFixture(block8.Header)
+			block9.Header.View = epoch2Setup.FinalView - safetyThreshold
+			block9.SetPayload(unittest.PayloadFixture(unittest.WithProtocolStateID(calculateExpectedStateId(block9.Header, nil))))
+
+			// TODO: check EpochExtension notification using pub/sub mechanism when it is implemented.
+
+			err = state.Extend(context.Background(), block9)
+			require.NoError(t, err)
+
+			epochProtocolState, err := state.AtBlockID(block9.ID()).EpochProtocolState()
+			require.NoError(t, err)
+			epochExtensions := epochProtocolState.Entry().CurrentEpoch.EpochExtensions
+			require.Len(t, epochExtensions, 1)
+			require.Equal(t, epochExtensions[0].FirstView, epoch2Setup.FinalView+1)
+
+			err = state.Finalize(context.Background(), block9.ID())
+			require.NoError(t, err)
+
+			// B10 will be the first block past the epoch extension
+			block10 := unittest.BlockWithParentProtocolState(block9)
+			block10.Header.View = epochExtensions[0].FirstView
+
+			unittest.InsertAndFinalize(t, state, block10)
+
+			// prepare epoch recover event for current epoch, we are using epoch extensions for current epoch
+			// since it has to be contiguous with the last extension.
+			epochRecover := unittest.EpochRecoverFixture(
+				unittest.WithParticipants(epoch2Participants),
+				unittest.SetupWithCounter(epoch2Setup.Counter+1),
+				unittest.WithFinalView(epochExtensions[0].FinalView+1000),
+				unittest.WithFirstView(epochExtensions[0].FinalView+1),
+			)
+
+			// B10 contains an EpochRecover event
+			receipt, seal = unittest.ReceiptAndSealForBlock(block10)
+			receipt.ExecutionResult.ServiceEvents = []flow.ServiceEvent{epochRecover.ServiceEvent()}
+			seal.ResultID = receipt.ExecutionResult.ID()
+
+			block11, epochRecoverSealingBlock := unittest.SealBlock(t, state, mutableProtocolState, block10, receipt, seal)
+
+			err = state.Finalize(context.Background(), block11.ID())
+			require.NoError(t, err)
+
+			// check on fork, invalid epoch transition attempted should be false
+			epochState, err := state.AtBlockID(epochRecoverSealingBlock.ID()).EpochProtocolState()
```

Comment:
I think `epochRecoverSealingBlock` is called block B12 in the test description? Would suggest to stick with `block12`